### ISSUE TITLE
about SFT

### ISSUE THREAD

**Issue by xuchanghua-88** [jd-opensource/JoyAI-VL-Interaction#22]
在llamafactory上进行sft时，请问能提供交叉熵损失加权计算的实现吗

**Comment by ydyhello** [comment]
> 在llamafactory上进行sft时，请问能提供交叉熵损失加权计算的实现吗

Below is a simplified version of the code. Streaming data samples use special loss weights.

```python
for sample in packed_batch:
    if sample.is_streaming:
        # Streaming data
        for token in sample.tokens:
            if token == "</response>":
                control_weight = response_token_weight
            elif token == "</silence>" and first_silence_after_response:
                control_weight = first_silence_token_weight
            elif token == "</silence>":
                control_weight = repeated_silence_token_weight
            else:
                control_weight = 1.0
    else:
        # Regular offline data
        control_weight = 1.0

    if label == -100:
        final_weight = 0.0
    else:
        final_weight = packing_weight * control_weight

    # All samples eventually enter Liger CE in the same way
    loss += CE(hidden_state[t], label[t + 1]) * final_weight[t + 1]
```
